base_action_agent/README.rst

```diff
@@ -0,0 +1,185 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+    :alt: License: AGPL-3
+
+Action Rules Extensions
+=======================
+
+Summary of features provided:
+
+  * Rule Sets organize Automated Rules in logical sets
+  * Run Automated Actions using a specific user.
+  * Define Rule filters composed of several combined Python snippets
+
+
+Rule sets are capable to provide an alternative workflow system:
+
+  * A Workflow is defined using a Rule Set
+  * Nodes, or States, are represented by Facts
+  * Signals, or transition conditions, are represented by Facts using
+    changed / old vs new values
+  * Transitions are represented by Action Rules
+
+
+With the "Run As" feature we have basic support for Bots or Agents:
+
+  * Create a User for the Bot. Make sure it has the necessary security permissions.
+  * Create a new Ruleset and set the Bot as its User.
+  * Create the Automated Action to be ran by the Bot.
+  * Check the Bot's ruleset "Enabled" flag to active the Rules.
+
+Make your OdooBot the best worker in the company!
+
+
+Rule Sets
+---------
+
+As the number of Action Rules increases it may get confusing.
+Rulesets allow to organize the rules per topic.
+
+Additionally you can assign a default User to a ruleset.
+Assign that to a bot User, and the ruleset becomes a bot playbook.
+You also have a "Enabled" flag to enable/disable the rules.
+
+The "Silence Error?" option allows for exceptions, other than
+``raise Warning()``, to be logged only and not propagated to the user.
+This is useful for some helper Bot use cases, where we don't want
+any Bot malfunction (missing security access, for example) to disturb
+the end user.
+
+
+Facts
+-----
+
+Facts are logic expressions to be evaluated.
+They can express state (ex: "Is Open") or events (iex: "Changed to Open").
+
+These are Python expressions, and their evaluation context has available:
+
+  * ``self``, ``obj``: is the record object, after the create/write operation
+  * ``env``, ``context``, and ``user``
+  * ``creating``, ``inserting``: True if triggered by a create operation
+  * ``writing``, ``updating``: True if triggered by a write operation
+  * ``old('<field>')``: gets the old value before the write operation
+  * ``new('<field>')``: gets the new value after the create/write operation
+  * ``changed('<field>')``: gets the new value, if changed or if inserting
```

Comment:
To me `changed` looks redundant with `new`, which is itself redundant with `self`.
Moreover, changing a field to a falsy value looks the same as not changing it, which I find confusing.
In the example below, `not changed('user_id')` actually means "responsible not changed **or changed to nobody**".

What about passing changes as a dictionary instead?
This could merge both `new` and `changed` in a less confusing API, IMHO.
Taking the example,
* `'user_id' not in changed` means "responsible not changed";
* `'user_id' in changed and not changed['user_id']` means "responsible changed to nobody".

We have incorporated similar ideas in version 9.0, but in a crappier form, though :confused: